pkg/backend/httpstate/policypack.go

```diff
@@ -174,24 +174,36 @@ func (pack *cloudPolicyPack) Publish(
 
 	fmt.Println("Compressing policy pack")
 
-	if runtime := op.PolicyPack.Runtime.Name(); !strings.EqualFold(runtime, "nodejs") {
-		return result.Errorf(
-			"failed to publish policies because Pulumi.yaml requests unsupported runtime %s",
-			runtime)
-	}
+	var packTarball []byte
 
 	// TODO[pulumi/pulumi#1307]: move to the language plugins so we don't have to hard code here.
-	packTarball, err := npm.Pack(op.PlugCtx.Pwd, os.Stderr)
-	if err != nil {
-		return result.FromError(
-			errors.Wrapf(err, "could not publish policies because of error running npm pack"))
+	runtime := op.PolicyPack.Runtime.Name()
+	if strings.EqualFold(runtime, "nodejs") {
+		packTarball, err = npm.Pack(op.PlugCtx.Pwd, os.Stderr)
+		if err != nil {
+			return result.FromError(
+				errors.Wrap(err, "could not publish policies because of error running npm pack"))
+		}
+	} else if strings.EqualFold(runtime, "python") {
+		// npm pack puts all the files in a "package" subdirectory inside the .tgz it produces, so we'll do
+		// the same for Python. That way, after unpacking, we can look for the PulumiPolicy.yaml inside the
+		// package directory to determine the runtime of the policy pack.
+		packTarball, err = archive.Tgz(op.PlugCtx.Pwd, "package", true /*useDefaultExcludes*/)
+		if err != nil {
+			return result.FromError(
+				errors.Wrap(err, "could not publish policies because of error creating the .tgz"))
+		}
+	} else {
+		return result.Errorf(
+			"failed to publish policies because PulumiPolicy.yaml specifies an unsupported runtime %s",
+			runtime)
```

Comment:
Not blocking - but seeing this does make me wonder - should we be delegating this logic to the plugins - so that new languages can be added without having to touch this code?  (not sure how far away we are from that being possible generally)

Reply:
> should we be delegating this logic to the plugins - so that new languages can be added without having to touch this code?

Ideally, yes. It'd involve defining a new analyzer gRPC method (e.g. `Package`) and then having the Policy SDK for each language provide an implementation.

For Python, this would mean porting the TGZ archiving implementation (with ignore behaviors) of this PR from Go to Python.

For Node.js, it'd mean calling `npm pack` from within Node.js Policy SDK. Though, we'd need to keep the existing Go implementation of shelling out to `npm pack` for previous versions of the Node.js Policy SDK where the `Package` gRPC method doesn't exist.

Reply:
Actually, I'm going to hold off on doing this for now. I agree it's the right long-term architecture to defer to the plugin to do this. We already have some related technical debt around installing dependencies for regular programs (which the TODO above attempts to point out, but should be linking to https://github.com/pulumi/pulumi/issues/1334, which I'll fix).